File: 0112-path-sum/0112-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        if root == None:
            return False
        
        stack = [root]
        while stack:
            curr = stack.pop()
            
            if curr.left:
                curr.left.val += curr.val
                stack.append(curr.left)
            if curr.right:
                curr.right.val += curr.val
                stack.append(curr.right)
                
            if not curr.left and not curr.right and curr.val == targetSum:
                return True
        
        return False

File: 0112-path-sum/0112-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int targetSum, int cumSum=0) {
        if (root == nullptr) {
            return false;
        }
        cumSum += root->val;
        if (root->left == nullptr && root->right == nullptr) {
            return (cumSum == targetSum);
        }
        bool lhs = hasPathSum(root->left, targetSum, cumSum);
        bool rhs = hasPathSum(root->right, targetSum, cumSum);
        
        return (lhs || rhs);
    }
};